bot.py

```diff
@@ -873,6 +906,64 @@ async def onVote(self, ctx):
         except Exception as ex:
             await self.bot.sendErrorToAuthor(ctx)
 
+
+    # Called periodically to try and clean up old games that aren't in progress anymore
+    # This is 8 hours, with inactivity at 3, as the expectation is that 3-4 hours is probably
+    # at the the common upper limit for a town to be active, so 5 hours should catch most games.
+    @tasks.loop(hours=8)
+    async def cleanupInactiveGames(self):
+        print("Checking for inactive games")
+        # Inactive games are ones that haven't had any activity in the last 3 hours
+        inactiveTime = datetime.datetime.now() - datetime.timedelta(hours=3)
+
+        numEnded = 0
+
+        townsToEndByGuild = {}
+        
+        # TODO: This could use the DB instead of a cog member
+
+        # Collect all the inactive towns
+        for g, go in self.currentlyActiveGames.items():
+            townsToEndByGuild[g] = []
+            for t, to in go.items():
+                if "lastActivity" not in to or to["lastActivity"] < inactiveTime:
+                    townsToEndByGuild[g].append(t)
+
+        # Actually run endGame on each inactive town
+        for g, ts in townsToEndByGuild.items():
+            guild = self.bot.get_guild(g)
+            if guild:
+                for t in ts:
+                    query = { "guild" : guild.id, "controlChannelId" : t }
+                    doc = guildInfo.find_one(query)
+                    if doc:
+                        info = TownInfo(guild, doc)
+                        try:
+                            await self.onEndGameInternal(guild, info)
+                            numEnded = numEnded + 1
+                        except Exception:
+                            pass
+
+        if numEnded > 0:
+            print("Ended " + str(numEnded) + " inactive game(s)")
+
+        # Remove all the inactive towns from our list
+        for g, ts in townsToEndByGuild.items():
+            for t in ts:
+                self.currentlyActiveGames[g].pop(t, None)
+            if not self.currentlyActiveGames[g]:
+                self.currentlyActiveGames.pop(g, None)
+
+        # If nothing is left active at all, stop the timer
+        if not self.currentlyActiveGames:
+            print("No remaining active games, stopping checks")
+            self.cleanupInactiveGames.stop()
+    
+    @cleanupInactiveGames.before_loop
```

Comment:
whoa what is this?

Reply:
It's some sort of attachment for timer functions - this gets called before the timer function is triggered, in this case to make sure we're actually connected to discord

Reply:
Apparently functions in python can have properties attached? I guess the timer system looks for this property on the function and if it exists it calls it first.